### PR TITLE
IGNITE-21132 Buffered disk writing when creating dump

### PR DESCRIPTION
**Pull request by yurinaryshkin**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/IgniteSnapshotManager.java [hunk 1]
```diff
@@ -421,6 +422,9 @@ public class IgniteSnapshotManager extends GridCacheSharedManagerAdapter
     /** Factory to working with delta as file storage. */
     private volatile FileIOFactory ioFactory = new RandomAccessFileIOFactory();
 
+    /** Factory to create dump. */
+    private volatile FileIOFactory dumpIoFactory = new WriteOnlyBufferedFileIOFactory();
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/IgniteSnapshotManager.java:426]: If we make `BufferedFileIO` decorator for any underlying `FileIO` we can push down `dumpIoFactory` to the dump create task and get rid of this changes.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/WriteOnlyBufferedFileIOFactory.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+ */
+public class WriteOnlyBufferedFileIOFactory implements FileIOFactory {
+    /** */
+    private static final long serialVersionUID = 0L;
+
+    /** */
+    public static final int DEFAULT_BUFFER_SIZE = 512 * 1024;
+
+    /** */
+    private final int bufSz;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/WriteOnlyBufferedFileIOFactory.java:37]: I think we must provide the way for the user to configure buffer size.

Comment (nizhikov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/WriteOnlyBufferedFileIOFactory.java:37]: Let's introduce `DistributedProperty` for it.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/CreateDumpFutureTask.java [hunk 3]
```diff
@@ -708,7 +714,12 @@ private boolean writtenByIterator(int cache, KeyCacheObject key, CacheObjectCont
             while (writers.get() > 0) // Waiting for all on the fly listeners to complete.
                 LockSupport.parkNanos(1_000_000);
 
-            U.closeQuiet(file);
+            try {
+                file.close();
+            }
+            catch (IOException e) {
+                throw new IgniteException(e);
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/CreateDumpFutureTask.java:721]: Do we have a test for that. The test need check that this exception doesn't affect Ignite node, only the ongoing dump process.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/BufferedFileIO.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+            return;
+
+        buf.flip();
+
+        if (buf.hasRemaining()) {
+            if (delegate.writeFully(buf) < 0)
+                throw new IOException("Couldn't write data");
+        }
+
+        buf.limit(buf.capacity()).position(0);
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/BufferedFileIO.java:119]: `buf.clear();`

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/BufferedFileIO.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    /** */
+    private void flush() throws IOException {
+        if (buf == null)
+            return;
+
+        buf.flip();
+
+        if (buf.hasRemaining()) {
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/BufferedFileIO.java:114]: Do we need this check?

**Review by timoninmaxim — CHANGES_REQUESTED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/CreateDumpFutureTask.java (hunk 3 above)
Comment (yurinaryshkin) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/CreateDumpFutureTask.java:721]: Yes, it is tested in `IgniteCacheDumpSelfTest#testDumpCancelOnFileCreateError`. Also added extra checks to make sure ignite provides data after the error.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/BufferedFileIO.java (hunk 5 above)
Comment (yurinaryshkin) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/BufferedFileIO.java:114]: Not needed. Removing the check.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/WriteOnlyZipFileIO.java [hunk 6]
```diff
@@ -17,142 +17,57 @@
...
-
-    /** {@inheritDoc} */
-    @Override public long size() throws IOException {
-        throw new UnsupportedOperationException();
-    }
-
-    /** {@inheritDoc} */
-    @Override public void clear() throws IOException {
-        throw new UnsupportedOperationException();
+        return len;
```
[the remark this answers is not in the document]
Comment (yurinaryshkin) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/WriteOnlyZipFileIO.java:64]: The result of WriteOnlyZipFileIO#write is used by writeFully method which counts how many bytes were consumed from the parameter srcBuf. If not all, then it invokes write() once again. So this method should return number of bytes consumed, not how many bytes were sent to delegate fileIO.

**Review by timoninmaxim — CHANGES_REQUESTED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/CreateDumpFutureTask.java [hunk 7]
```diff
@@ -173,7 +172,14 @@ public CreateDumpFutureTask(
         );
 
         this.dumpDir = dumpDir;
-        this.ioFactory = compress ? new WriteOnlyZipFileIOFactory() : ioFactory;
+
+        ioFactory = new BufferedFileIOFactory(ioFactory);
+
+        if (compress)
+            ioFactory = new WriteOnlyZipFileIOFactory((BufferedFileIOFactory)ioFactory);
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/CreateDumpFutureTask.java:179]: Useless cast

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/BufferedFileIO.java [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+        buf.put(b);
+
+        if (!buf.hasRemaining())
+            flush();
+    }
+
+    /** */
+    private void flush() throws IOException {
+        if (buf == null)
+            return;
```
[the remark this answers is not in the document]
Comment (yurinaryshkin) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/BufferedFileIO.java:110]: This is to allow double close() invocation. The method is private that's why its behavior is slightly different than other public methods.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/BufferedFileIO.java [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+                flush();
+
+            srcBuf.limit(limit);
+        }
+
+        return bytesCnt;
+    }
+
+    /** {@inheritDoc} */
+    @Override public int write(byte[] srcBuf, int off, int len) throws IOException {
```
[the remark this answers is not in the document]
Comment (yurinaryshkin) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/BufferedFileIO.java:70]: Yes. This approach will simplify code, but will not allow to use DirectByteBuffer as a source of data. Will unify for now and once DirectBB will be used, we can revert the Buffer solution.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/BufferedFileIO.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (yurinaryshkin) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/BufferedFileIO.java:110]: Extracted the check into close() - the only place where the check is needed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/BufferedFileIOFactory.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    /** {@inheritDoc} */
+    @Override public BufferedFileIO create(File file, OpenOption... modes) throws IOException {
+        FileIO io = factory.create(file, modes);
+
+        int blockSize = io.getFileSystemBlockSize();
+
+        if (blockSize < 0)
+            blockSize = DEFAULT_BLOCK_SIZE;
+
+        return new BufferedFileIO(io, blockSize);
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/BufferedFileIOFactory.java:53]: Looks like that blockSize can be calculated within BufferedFileIO

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/WriteOnlyZipFileIO.java [hunk 11]
```diff
@@ -17,142 +17,57 @@
...
 import static java.util.zip.Deflater.BEST_COMPRESSION;
-import static org.apache.ignite.internal.processors.cache.persistence.file.FilePageStoreManager.ZIP_SUFFIX;
 
 /**
  * {@link FileIO} that allows to write ZIP compressed file.
  * It doesn't support reading or random access.
  * It is not designed for writing concurrently from several threads.
  */
-public class WriteOnlyZipFileIO extends AbstractFileIO {
+public class WriteOnlyZipFileIO extends FileIODecorator {
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/WriteOnlyZipFileIO.java:35]: Let's inherit it from BufferedFileIO instead. It makes this class usage much simpler.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/CreateDumpFutureTask.java [hunk 12]
```diff
@@ -173,7 +172,14 @@ public CreateDumpFutureTask(
         );
 
         this.dumpDir = dumpDir;
-        this.ioFactory = compress ? new WriteOnlyZipFileIOFactory() : ioFactory;
+
+        ioFactory = new BufferedFileIOFactory(ioFactory);
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/CreateDumpFutureTask.java:176]: Let's change type of ioFactory to BufferedFileIOFactory.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/CreateDumpFutureTask.java [hunk 13]
```diff
@@ -644,7 +650,7 @@ private void write(
 
             rateLimiter.acquire(buf.limit());
 
-            if (file.writeFully(buf) != buf.limit())
+            if (file.write(buf) != buf.limit())
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/CreateDumpFutureTask.java:653]: Let's make file type BufferedFileIO and override writeFully method instead

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/BufferedFileIO.java [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+ */
+public class BufferedFileIO extends FileIODecorator {
+    /** */
+    private static final int DEFAULT_BLOCK_SIZE = 4096;
+
+    /** */
+    private ByteBuffer buf;
+
+    /** */
+    private static int bufSize(FileIO io) {
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/dump/BufferedFileIO.java:40]: Let's inline this method

**Review by timoninmaxim — APPROVED** [review]

**Pull request merged**